custom-types-quick-guide.md

```diff
@@ -0,0 +1,245 @@
+# Custom Types Quick Guide
+
+_Abstract._ We will go over the steps required to extend an existing service with a new API. The
+focus is on covering the details with respect to the client protocol and its related components. We
+will not cover the underlying logic of the API itself. We use Java as the subject language for code
+generation. The [README](README.md) covers the general framework, so please read that first.
+
+## Adding a Service API
+
+Let us take the `DynamicConfig.yaml` service as the baseline for an example. We will add a
+fictitious method: `addUsefulConfig`.
+
+_Note_. At the time of writing the highest `id` in `DynamicConfig.yaml` was `18` -- the `id` of the
+last entry, hence the use of `19` for our example.
+
+```yaml
+- id: 19
+  name: addUsefulConfig
+  since: 2.7
+  doc: |
+    Adds a useful configuration.
+  request:
+    retryable: false
+    partitionIdentifier: -1
+    params:
+      - name: name
+        type: String
+        nullable: false
+        since: 2.7
+        doc: |
+          The configuration name.
+      - name: serviceAConfig
+        type: ServiceAConfig
+        nullable: false
+        since: 2.7
+        doc: |
+          Configuration for Service A.
+      - name: serviceBConfig
+        type: ServiceBConfig
+        nullable: false
+        since: 2.7
+        doc: |
+          Configuration for Service B.
+  response: {}
+```
+
+We have 3 fields in this new `addUsefulConfig` API:
+
+- `name` which is a `String`. This is trivial so we won't touch on this again.
+- `serviceAConfig` which is a `ServiceAConfig` -- a custom type we will create shortly.
+- `serviceBConfig` which is a `ServiceBConfig` -- a custom type we will create shortly.
+
+We have two custom types here: `ServiceAConfig` will model a Java configuration type that is defined
+by standard language types, e.g. `int`, `String`, etc. By contrast, `ServiceBConfig` will be defined
+by framework specific types that don't have a direct mapping in other target languages.
+
+If we run `./generator.py` we will get some `dict` lookup errors because we've referenced some
+custom types. In `Custom.yaml` we'll append the following:
+
+```yaml
+- name: ServiceAConfig
+  since: 2.7
+  params:
+    - name: address
+      type: String
+      nullable: false
+      since: 2.7
+    - name: port
+      type: int
+      nullable: false
+      since: 2.7
+- name: ServiceBConfig
+  since: 2.7
+  params:
+    - name: address
+      type: String
+      nullable: false
+      since: 2.7
+    - name: addressParser
+      type: Data
+      nullable: false
+      since: 2.7
+```
+
+Open `schema/protocol-schema.json` and append to the array `definitions.param.properties.type.enum`
+the string literals: `ServiceAConfig` and `ServiceBConfig`.
```

Comment:
I think if we have used a new custom type in a custom type, we would also need to modify custom-codec-schema.json. 

Should we mention it here as well?

Reply:
ok, suggested just to make it comprehensive but not a must